driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java

```diff
@@ -107,30 +111,62 @@ public final class CommandMessage extends RequestMessage {
         this.serverApi = serverApi;
     }
 
+    /**
+     * Create a BsonDocument representing the logical document encoded by an OP_MSG.
+     * <p>
+     * The returned document will contain all the fields from the Body (Kind 0) Section, as well as all fields represented by
+     * OP_MSG Document Sequence (Kind 1) Sections.
+     */
     BsonDocument getCommandDocument(final ByteBufferBsonOutput bsonOutput) {
-        ByteBufBsonDocument byteBufBsonDocument = ByteBufBsonDocument.createOne(bsonOutput,
-                getEncodingMetadata().getFirstDocumentPosition());
-        BsonDocument commandBsonDocument;
-
-        if (containsPayload()) {
-            commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
-
-            int payloadStartPosition = getEncodingMetadata().getFirstDocumentPosition()
-                    + byteBufBsonDocument.getSizeInBytes()
-                    + 1 // payload type
-                    + 4 // payload size
-                    + payload.getPayloadName().getBytes(StandardCharsets.UTF_8).length + 1;  // null-terminated UTF-8 payload name
-            commandBsonDocument.append(payload.getPayloadName(),
-                    new BsonArray(ByteBufBsonDocument.createList(bsonOutput, payloadStartPosition)));
-        } else {
-            commandBsonDocument = byteBufBsonDocument;
+        List<ByteBuf> byteBuffers = bsonOutput.getByteBuffers();
+        try {
+            CompositeByteBuf byteBuf = new CompositeByteBuf(byteBuffers);
+            try {
+                byteBuf.position(getEncodingMetadata().getFirstDocumentPosition());
+                ByteBufBsonDocument byteBufBsonDocument = createOne(byteBuf);
+
+                // If true, it means there is at least one Kind 1:Document Sequence in the OP_MSG
+                if (byteBuf.hasRemaining()) {
+                    BsonDocument commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
+
+                    // Each loop iteration processes one Document Sequence
+                    // When there are no more bytes remaining, there are no more Document Sequences
+                    while (byteBuf.hasRemaining()) {
+                        byteBuf.position(byteBuf.position() + 1 /* payload type */ + 4 /* payload size */);
+                        String payloadName = getPayloadName(byteBuf);
+                        assertFalse(payloadName.contains("."));
+                        commandBsonDocument.append(payloadName, new BsonArray(createList(byteBuf)));
```

Comment:
`createList(byteBuf)` never stops, until it reaches the end of `byteBuf`. That is a problem, given that we are trying to decode multiple sections, as opposed to decoding a single section. Given that we expect the payload of all sections to be sequences, we need to read the sequence size and use `createList` to decode the documents in the sequence only up to the position where the sequence ends.

Unfortunately, given that `CommandMessage` can't currently have multiple `SplittablePayload`s, we can neither test the PR code, nor the proposed fix for it. 

Following is how I propose to fix the code:

```suggestion
                        // skip reading the payload type, we know it is 1
                        byteBuf.position(byteBuf.position() + 1);
                        int sequenceStart = byteBuf.position();
                        int sequenceSizeInBytes = byteBuf.getInt();
                        int sectionEnd = sequenceStart + sequenceSizeInBytes;
                        String sequenceIdentifier = getSequenceIdentifier(byteBuf);
                        assertFalse(sequenceIdentifier.contains("."));
                        ByteBuf documentsByteBufSlice = byteBuf.duplicate().limit(sectionEnd);
                        byteBuf.position(sectionEnd);
                        commandBsonDocument.append(sequenceIdentifier, new BsonArray(createList(documentsByteBufSlice)));
```

Reply:
I'm going to grab this suggestion and commit it locally, since I've made further changes that conflict.

Reply:
I basically used this as is, except I added a release of `documentsByteBufSlice` in a finally block.

Thanks for find this.  It was careless of me.

Reply:
> I added a release of `documentsByteBufSlice`

🤦‍♂️ Thank you for spotting that)